Sources/Algorithms/PartialSort.swift

```diff
@@ -7,280 +7,80 @@
-  /// Returns the elements of the sequence such that the 0...k range contains
-  /// the first k smallest elements in this sequence.
-  ///
-  /// This example partially sorts an array of integers to retrieve its three
-  /// smallest values:
-  ///
-  ///     let numbers = [7,1,6,2,8,3,9]
-  ///     let almostSorted = numbers.partiallySorted(3)
-  ///     // [1, 2, 3, 9, 7, 6, 8]
-  ///     let smallestThree = almostSorted.prefix(3)
-  ///     // [1, 2, 3]
-  ///
-  /// The order of equal elements is not guaranteed to be preserved, and the
-  /// order of the remaining elements is unspecified.
-  ///
-  /// If you need to sort a sequence but only need access to a prefix of
-  /// its elements, using this method can give you a performance boost over
-  /// sorting the entire sequence.
-  ///
-  /// - Parameter count: The k number of elements to partially sort
-  /// in ascending order.
-  ///
-  /// - Complexity: O(k log n)
-  public func partiallySorted(_ count: Int) -> [Element] {
-    return partiallySorted(count, by: <)
-  }
-}
+    // If we're attempting to prefix more than 10% of the collection, it's faster to sort everything.
+    guard count < (self.count / 10) else {
+      return Array(try sorted(by: areInIncreasingOrder).prefix(count))
+    }
 
-extension MutableCollection where Self: RandomAccessCollection {
-  /// Rearranges this collection such that the 0...k range contains the first
-  /// k sorted elements in this collection, using the given predicate as the
-  /// comparison between elements.
-  ///
-  /// This example partially sorts an array of integers to retrieve its three
-  /// smallest values:
-  ///
-  ///     var numbers = [7,1,6,2,8,3,9]
-  ///     numbers.partiallySort(3, <)
-  ///     // [1, 2, 3, 9, 7, 6, 8]
-  ///     let smallestThree = numbers.prefix(3)
-  ///     // [1, 2, 3]
-  ///
-  /// The order of equal elements is not guaranteed to be preserved, and the
-  /// order of the remaining elements is unspecified.
-  ///
-  /// If you need to sort a collection but only need access to a prefix of its
-  /// elements, using this method can give you a performance boost over sorting
-  /// the entire collection.
-  ///
-  /// - Parameter count: The k number of elements to partially sort.
-  /// - Parameter areInIncreasingOrder: A predicate that returns true if its
-  /// first argument should be ordered before its second argument;
-  /// otherwise, false.
-  ///
-  /// - Complexity: O(k log n)
-  public mutating func partiallySort(
-    _ count: Int,
-    by areInIncreasingOrder: (Element, Element) throws -> Bool
-  ) rethrows {
-    try __partiallySort(count, by: areInIncreasingOrder)
+    var result = try self.prefix(count).sorted(by: areInIncreasingOrder)
+    for e in self.dropFirst(count) {
+      if let last = result.last, try areInIncreasingOrder(last, e) { continue }
+      if let insertionIndex = try result.firstIndex(where: { try areInIncreasingOrder(e, $0) }) {
```

Comment:
Does the `firstIndex` here outperform `partitioningIndex`? It won't change the time complexity either way, but if they perform equally well then it might be worth minimizing the number of calls to `areInIncreasingOrder` (in case it is computationally intensive).

Reply:
No, it makes sense to use `partitioningIndex`. As an additional optimization I also made it remove the last element _before_ trying to insert the new one to prevent shifting an element without necessity.